Propagate headers/trailers if response isn't gRPC

ejona86 opened the pull request:
This provides more structured data into the application for it to do
special handling.

In general, we would hope most people don't need this functionality, but
it is a good escape hatch to allow users to workaround infrastructure
problems.

I do think there needs to be some discussion on whether we think this is
a good idea, since currently we are only delivering metadata if it was sent
from a gRPC client/server.

ejona86 commented:
I discussed with some other devs, and there weren't major concerns. Still some people in the wider grpc team to discuss with, but in general it isn't viewed as outrageous.

carl-mastrangelo commented:
As a temporary solution I am okay with this.  However, proper verbiage about expectations should be added to make sure users of onClose know (e.g. via javadoc) that the metadata provided may or may not be populated in the future.  Having the bad headers propagate up this high in the stack doesn't intuitively feel like a good solution.

ejona86 commented:
@carl-mastrangelo, I added OkHttp tests; before I didn't add any because I didn't find some that were similar, but I found some this time. I also added a test for initial HEADERS with EOS (like would be received if there isn't an HTTP entity) because there wasn't one verifying that.

carl-mastrangelo commented on core/src/main/java/io/grpc/internal/AbstractClientStream.java:
```diff
@@ -95,7 +106,7 @@ protected void inboundTransportError(Status errorStatus) {
+    transportReportStatus(errorStatus, false, metadata);
   }
 
   /**
```
Since I can't comment on lines you didn't add:  Until the new compression change is in, you should probably also change line 132 to include the headers.

ejona86 replied:
Done, with verification in test. That left only one call-site for the 1-arg version of transportReportStatus, so I converted it to the 2-arg version and removed the 1-arg version of the method.

carl-mastrangelo commented:
@ejona86  One minor comment, but LGTM

ejona86 commented:
@carl-mastrangelo, PTAL
